### PR TITLE
Fix type hint for dotenv_path var, add StrPath alias

### LINKED ISSUE #431: pyright in strict mode errors on `from dotenv import load_dotenv`
Steps to reproduce:

```bash
$ python --version
Python 3.9.13
$ mkdir repro
$ cd repro
$ cat > repro.py <<EOL
from dotenv import load_dotenv
load_dotenv()
EOL
$ echo '{ "typeCheckingMode": "strict" }' > pyrightconfig.json
$ python -m venv .venv
$ source .venv/bin/activate
$ pip install pyright python-dotenv
Collecting pyright
  Using cached pyright-1.1.276-py3-none-any.whl (16 kB)
Collecting python-dotenv
  Using cached python_dotenv-0.21.0-py3-none-any.whl (18 kB)
Collecting nodeenv>=1.6.0
  Using cached nodeenv-1.7.0-py2.py3-none-any.whl (21 kB)
Requirement already satisfied: setuptools in ./.venv/lib/python3.10/site-packages (from nodeenv>=1.6.0->pyright) (58.1.0)
Installing collected packages: python-dotenv, nodeenv, pyright
Successfully installed nodeenv-1.7.0 pyright-1.1.276 python-dotenv-0.21.0
$ pyright repro.py
Loading configuration file at /Users/eaftan/repro/pyrightconfig.json
Assuming Python version 3.9
Assuming Python platform Darwin
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /Users/eaftan/repro/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.276
/Users/eaftan/repro/repro.py
  /Users/eaftan/repro/repro.py:1:20 - error: Type of "load_dotenv" is partially unknown
    Type of "load_dotenv" is "(dotenv_path: str | PathLike[Unknown] | None = None, stream: IO[str] | None = None, verbose: bool = False, override: bool = False, interpolate: bool = True, encoding: str | None = "utf-8") -> bool" (reportUnknownVariableType)
1 error, 0 warnings, 0 informations 
Completed in 0.579sec
```

I think the problem is that [`load_dotenv` is typed as follows](https://github.com/theskumar/python-dotenv/blob/cedd36d395b772adb556ba4635ab42d59ad27567/src/dotenv/main.py#L307):

```python
def load_dotenv(
    dotenv_path: Union[str, os.PathLike, None] = None,
    stream: Optional[IO[str]] = None,
    verbose: bool = False,
    override: bool = False,
    interpolate: bool = True,
    encoding: Optional[str] = "utf-8",
) -> bool:
```

`os.PathLike` is a generic type and should have a type argument.  I think the correct typing for `dotenv_path` should be `Union[str, os.PathLike[str], os.PathLike[bytes], None]`.

### PR DESCRIPTION
**Pull request by eaftan**
Fixes #431

### CODE REVIEW

File: src/dotenv/main.py [hunk 1]
```diff
@@ -305,7 +305,7 @@ def _is_interactive():
 
 
 def load_dotenv(
-    dotenv_path: Union[str, os.PathLike, None] = None,
+    dotenv_path: Union[str, 'os.PathLike[str]', 'os.PathLike[bytes]', None] = None,
```
Comment (joe733) [inline src/dotenv/main.py:308]: Is this better: `Union[str, os.PathLike[Union[str, bytes]], None]`?

Comment (joe733) [reply src/dotenv/main.py:308]:
```python
dotenv_path: str | os.PathLike[str | bytes] | None = None
```

image

is compact and nice, but currently [supported from Python 3.10 and above](https://peps.python.org/pep-0604/). So yes [typing.Optional](https://docs.python.org/3/library/typing.html#typing.Optional) would also be fine.

```python
from typing import Optional, Union
...
dotenv_path: Optional[Union[str, os.PathLike[Union[str, bytes]]] = None
```

**Comment by KalleDK** [comment]
Anything needed to make this pr happen? :) Came here for this exact problem :)

**Comment by Viicos** [comment]
@eaftan I think a type alias could be used, as the `dotenv_path` argument is also present in other places in the code, and this PR could also fix the type hint for these ones

**Comment by eaftan** [comment]
> @eaftan I think a type alias could be used, as the `dotenv_path` argument is also present in other places in the code, and this PR could also fix the type hint for these ones

Done.  The CI seems to be failing with a coverage error, but I think that's not my fault -- it seems to have failed on the most recent commit to main as well: https://github.com/github/entitlements/pull/56799

**Comment by theskumar** [comment]
Thank you @eaftan @Viicos!

 I'll merge and look into the CI issues.

**Pull request merged**